### ISSUE TITLE
What makes CropTarget special to require an asynchronous creation?

### ISSUE THREAD

**Issue by youennf** [w3c/mediacapture-region#17]
We started discussing this topic (whether CropTarget creation should use promise or not) as part of https://github.com/w3c/mediacapture-region/issues/11 and it seems best to use a dedicated thread for this issue.

@eladalon1983  mentioned implementation issues that make using promises desirable, in particular security and potential race conditions if creating synchronously a CropTarget since CropTarget has to keep a link to the process it was created in. The following case was provided:
- create CropTarget in process A
- postMessage it synchronously to another process B
- (optionally) postMessage it yet again to another process C
- Use CropTarget to crop the track.
@eladalon1983, please correct this description if it is not aligned with your thoughts.

Existing objects like MessagePort, WHATWG Streams, RTCDataChannel or MediaStreamTrack can be created-then-postMessaged synchronously and UAs are implementing this today, hopefully without the security/race condition issues.
AIUI, it seems consistent to use the same approach to CropTarget (synchronous creation), except if we find anything specific to CropTarget that actually prevents this existing pattern.

**Comment by youennf** [comment]
When transferring a MessagePort or a RTCDataChannel from SLIDE to VC, there is a need to identify that the transferred object is originating from SLIDE, just like CropTarget. How is it possible to be synchronous for those objects but not for CropTarget?

**Comment by youennf** [comment]
Again, the same could be said about transferring a RTCDataChannel from SLIDE->TOP_LEVEL->VC.
Either this is a new security hardening rule that we cannot enforce it in old APIs but should in new APIs or there is something specific to CropTarget. I am unclear which one it is from your explanations.

**Comment by youennf** [comment]
What about MessagePort then? MessagePort are very common and designed specifically for being transferred.
From what I can see so far, the problems you are describing have been solved for those objects in a secure and efficient way without using promises.

**Comment by eladalon1983** [comment]
Does `MessagePort` remain bound in any way, or encode to some degree, the document in which it was originally instantiated? If I create a `MessagePort` in D1, then transfer it to D2, then to D3, will D3 know that this `MessagePort` is originally from D1?

**Comment by youennf** [comment]
> Does `MessagePort` remain bound in any way, or encode to some degree, the document in which it was originally instantiated?

It remains bound to the MessagePort it was created jointly through MessageChannel.

> If I create a `MessagePort` in D1, then transfer it to D2, then to D3, will D3 know that this `MessagePort` is originally from D1?

For RTCDataChannel, WHATWG Streams and MediaStreamTrack yes, always, the 'source' remains in D1.

MessagePort are a bit specific in that they are created as a pair through MessageChannel that can keep communicating with each other. MessageChannel.port1 can be synchronously transferred to another process, as well as MessageChannel.port2. port1 and port2 remain bound together in any case.

**Comment by youennf** [comment]
Another example that might be closer to CropTarget is OffscreenCanvas.
OffscreenCanvas can be created from a canvas element, in which case it remains tied to the canvas element like CropTarget would be to its element. Such OffscreenCanvas can be created-then-transferred synchronously.

**Comment by jan-ivar** [comment]
@eladalon1983 said in https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1023100726:
> - When D2 calls cropTo(X), the UA has to validate that X is a valid CropTarget.
> - It is undesirable to query all documents and check if any of them have produced X

[cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto) already returns a promise, so querying "all" documents in the single viewport captured to identify the crop target, seems reasonable to me.

Cost to implementers is low on the [priority of constituencies](https://www.w3.org/TR/html-design-principles/#priority-of-constituencies) when it comes to shaping APIs.

**Comment by eladalon1983** [comment]
> [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto) already returns a promise, so querying "all" documents in the single viewport captured to identify the crop target, seems reasonable to me.

I think you mean "browsing context" where you said "viewport".
Do you have a reasonable limit on how many different documents could be embedded in that browsing context?

> Cost to implementers is low on the [priority of constituencies](https://www.w3.org/TR/html-design-principles/#priority-of-constituencies) when it comes to shaping APIs.

Cost to implementers is low priority, but non-zero. It's only a problem if accommodating implementers comes at a non-trivial cost to a higher-priority constituency. Does it?

**Comment by jan-ivar** [comment]
> I think you mean "browsing context" where you said "viewport".

No, each iframe has its own [browsing context](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context), which is nested (immediately or multiple levels under) the [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context).

I loosely mean all documents in this capture, which maybe translates to the top-level browsing context's document and all documents in its [nested browsing context](https://html.spec.whatwg.org/#nested-browsing-context)s of iframes that [intersect the viewport](https://html.spec.whatwg.org/#intersect-the-viewport).

> Do you have a reasonable limit on how many different documents could be embedded in that browsing context?

Looking up the CropTarget shouldn't be the bottleneck in extreme cases, so this should scale fine.

**Comment by eladalon1983** [comment]
> > I think you mean "browsing context" where you said "viewport".
> 
> No, each iframe has its own [browsing context](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context), which is nested (immediately or multiple levels under) the [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context).
> 
> I loosely mean all documents in this capture, which maybe translates to the top-level browsing context's document and all documents in its [nested browsing context](https://html.spec.whatwg.org/#nested-browsing-context)s of iframes that [intersect the viewport](https://html.spec.whatwg.org/#intersect-the-viewport).

I've been carrying that mistake around for a while. Thanks for enlightening me.

> > Do you have a reasonable limit on how many different documents could be embedded in that browsing context?
> 
> Looking up the CropTarget shouldn't be the bottleneck in extreme cases, so this should scale fine.

IPC with multiple processes is neither simple, nor performant, nor robust. The cost to implementers is greatly reduced when avoiding this. What's the downside to any other constituency?

**Comment by youennf** [comment]
> IPC with multiple processes is neither simple, nor performant, nor robust. The cost to implementers is greatly reduced when avoiding this.

This is a known problem that is solved in modern browsers. A transferred WritableStream should not do multiple IPCes to locate the process of its sink when writing new values.
As I said before, I'd like to understand why it would be more difficult with CropTarget than with all these other existing APIs.

> What's the downside to any other constituency?

It is more costly to both web developers and web engines. It is not consistent with existing Web APIs AFAIK.

**Comment by youennf** [comment]
Given this is a solved problem for other APIs and given these solutions are applicable to CropTarget as well, can we converge on moving away from Promises?

**Comment by eladalon1983** [comment]
> > IPC with multiple processes is neither simple, nor performant, nor robust. The cost to implementers is greatly reduced when avoiding this.
> 
> This is a known problem that is solved in modern browsers. A transferred WritableStream should not do multiple IPCes to locate the process of its sink when writing new values. As I said before, I'd like to understand why it would be more difficult with CropTarget than with all these other existing APIs.

I believe I have explained why we have implemented things this way in Chrome. This is a real issue.


> > What's the downside to any other constituency?
> 
> It is more costly to both web developers and web engines. It is not consistent with existing Web APIs AFAIK.

The cost to Web developers is negligible. Crop-target production is a rare occurrence, it does not matter to the Web developer if it complete asynchronously. I can pull in Web-developers currently using Region Capture (in origin trial) for major products with a high level of polish, and they could comment as much. Would you find that convincing? (If not - please pull in a similarly qualified Web developer who could comment to the contrary.)

> Given this is a solved problem for other APIs and given these solutions are applicable to CropTarget as well, can we converge on moving away from Promises?

Let's converge **towards** Promises, given that it's an important implementation issue for Chrome. (And I believe that when the time comes for Safari and Firefox to implement this, they'll find it equally problematic.)

**Comment by youennf** [comment]
> I believe I have explained why we have implemented things this way in Chrome. This is a real issue.

You explained a real issue, that I would classify as an optimization problem (though at some point you alluded to security concerns as well). Is that correct?

The argument is that the same optimization issue exists for already deployed APIs, and was solved without making use of promises. If we do not want to follow this preexisting pattern, we need a clear justification.

To move forward, let's try with more narrowly focused questions:
- Do you think that other existing web APIs enumerated in this thread have the same issue? Or do you think this is new problem specific to region?
-  If it is a new problem, can you detail what exactly is new?
- If it is not a problem specific to region, what makes you think the solutions used for implementing those APIs are not good enough for region?

As a recap, here are some APIs that I think face the same issue:
- Create a MessageChannel and transfer one of its MessagePort synchronously after creation. The transferred port needs to be able to locate in an optimal manner where is living its related channel port.
- Create a MediaStreamTrack from a canvas and transfer it synchronously after creation. The transferred track needs to be able to locate in an optimal manner where is living the canvas element.
- Create a RTCDataChannel from a RTCPeerConnection and transfer it synchronously after creation. The transferred data channel needs to be able to locate in an optimal manner where is living the peer connection.

> I believe that when the time comes for Safari and Firefox to implement this, they'll find it equally problematic.

I implemented in WebKit some of the APIs that I think are facing the issue you are describing.
For that purpose, existing techniques were reused so that we do not introducing delay in creation of the objects/delay in transferring the objects/delay in the object algorithms/race conditions.

Answering the above questions might help figuring out which problems I might be overlooking.

**Comment by alvestrand** [comment]
Note: When we started trying to transfer MediaStreamTracks in Chrome, the synchronous nature of the transfer gave us major problems in implementation. So the idea that synchronous = solved problem is not universally true.

**Comment by youennf** [comment]
It is great to hear Chrome implementation on transferring tracks is making progress. It is also great to hear Chrome implemented the transfer of synchronously-created MediaStreamTracks in a secure and efficient manner.
Can you validate whether the same approach can be used for CropTarget?

> The idea that synchronous = solved problem is not universally true.

The point is more about create-then-transfer-synchronously is a solvable problem (do we all agree?), it was solved multiple times already.
And that the web platform never cared about this particular implementation difficulty when designing new APIs.

To break this existing pattern, compelling motivations seem necessary.

Another reason not to use promises: what happens in case the element goes transferred to another document, which then gets destroyed (and the element gets reattached to another document), all of this during creation of the CropTarget. Should we reject the promise? A synchronous API is simpler for edge cases as well as for web developers.

**Comment by eladalon1983** [comment]
> what happens in case the element goes transferred to another document

I've had the same concern. I lost that concern when I learned that Elements are not transferable. (But do correct me if I am wrong.)

> which then gets destroyed

It is generally possible for a CropTarget to outlive its Element, and that's OK. The document discusses what happens then. The summary is:
* If cropped to a target that is in the DOM, has >0 pixels, etc., frames are produced.
* Otherwise, no further frames are produced. The track is effectively "muted" (but without the muting - as per our earlier agreement).
* Elements can be garbaged collected even! That's not special. It falls into the category set by the previous bullet-point.
* It's always possible to get a track "unstuck" by calling cropTo again, either to uncrop it or to re-crop it to a new target.

**Comment by youennf** [comment]
> I had the same concern. I lost that concern when I learned that Elements are not transferable.

The point I am making is during the creation of the CropTarget, i.e. when the promise is not yet settled.

**Comment by eladalon1983** [comment]
> > I had the same concern. I lost that concern when I learned that Elements are not transferable.
> 
> The point I am making is during the creation of the CropTarget, i.e. when the promise is not yet settled.

The thing I am still not getting is - what's going to happen to the Element during that time? The worst that could happen is that it gets garbage collected. I don't think that's a problem. It doesn't seem to matter if the Element is GCed before/after its CropTarget is produced. (And getting GCed after CropTarget-production should be a normal occurrence.)

**Comment by jan-ivar** [comment]
I should have posted https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1126528615 here. To summarize it: At least two highly skilled technical people were confused by the current API into thinking it does more than it does.

That's a cost to web developers that we should and do avoid on the regular, as @youennf shows.

**Comment by jan-ivar** [comment]
> (Minting can fail if the browser process refuses to record more mappings.)

This is an incorrect implementation since [produceCropTarget](https://w3c.github.io/mediacapture-region/#producecroptarget-method) is infallible.

**Comment by eladalon1983** [comment]
> I should have posted [#11 (comment)](https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1126528615) here. To summarize it: At least two highly skilled technical people were confused by the current API into thinking it does more than it does.

It's unclear to me who was confused and in what way, and how making the API synchronous would solve their confusion.

**Comment by youennf** [comment]
I would first like to get consensus on one API design point. Hopefully, we can all agree on something like:
1. For any new API, we try to make it synchronous if we can.
2. If the algorithm requires some hopping to other threads/processes/environments, we switch to async.
3. If synchronous implementations are overly complex, we switch to async.

AIUI, Chrome current implementation is asynchronous, and @eladalon1983 is stating that doing a synchronous implementation is overly complex.
@jan-ivar and @youennf think that such a synchronous implementation is not complex. Some implementation strategies and already implemented APIs that deal with the same issue have been suggested.
The rest of the message (sorry for its length) is describing that.

> * **MessageChannel:**
>   
>   * If implementing with direct communication between the processes, the risk involved is a necessary evil. This cannot be said for CropTarget. Discoverability in either direction is not a requirement here, and confers little to no benefit.
>   * If implementing with mediation via another process, the story gets more complicated. A valid implementation can hide that it's asynchronically minting identifiers, behind the moment of posting the MessageChannel to the other document. (Some compromises are required, though.) But I don't want to discuss this because it would lose track of the topic - see below.

@eladalon1983, I think we do agree MessagePort can be created and transferred cross process synchronously.
Can you validate Chrome implementation of MessagePort creation/transfer, in particular whether Chrome is minting asynchronously identifiers for MessagePorts?

FWIW, given we now have such MessagePort, we can reuse MessagePort to implement CropTarget.
Below algorithm is taking the hypothesis that CropTarget is transferable and can be used only once (these limitations can be easily fixed by recreating MessageChannels as required):
1. At CropTarget creation, create a MessageChannel, set port1 to a slot of the element, and port2 to a CropTarget slot.
2. When transferring CropTarget to another environment, transfer CropTarget.port2, and recreate a new CropTarget with the transferred port2
3. When calling cropTo with the new CropTarget, transfer CropTarget.port2 to the process doing the capture.
4. In the capture process, use port2 to communicate with the element (through port1) to gather the necessary states to start the actual cropping. 

Steps 1 and 2 can be done synchronously.
Steps 3 and 4 are asynchronous which is fine since they are run as part of cropTo which returns a promise.

> * **MediaStreamTrack:**
>   
>   * These are **not** currently transferrable in Chrome - not synchronously, not at all. To claim it's possible one needs to present an implementation  as proof, not a specification. (Does it work on Safari or Firefox yet? [This demo](https://cheerful-zany-editorial.glitch.me/) suggests that it's not, as of 2022-03-31.)
>   * My colleauges working on mst-transferability tell me that they are running into **a lot** of issues **precisely** because of the requirement that they be "synchronously transferable".

It would be nice to hear about the exact MST issues.
I would bet this is due to MST complex state handling.
CropTarget has no changing state, which makes it a much simpler object to transfer.
That said, video MST are probably already transferable synchronously in Chrome by doing the following:
1. Create a video MediaStreamTrack from canvas
2. Get the ReadableStream from MediaStreamTrack using MediaStreamTrackProcessor
3. Transfer the ReadableStream
4. Recreate the MediaStreamTrack from the transferred ReadableStream using VideoTrackGenerator

Again step 1, 2 and 3 are all synchronous, from creation of a MediaStreamTrack to transferring the MediaStreamTrack via ReadableStream.

> * **RTCDataChannel:**
>   
>   * As of the time of this writing, these don't appear to be transferable in neither [the specification](https://w3c.github.io/webrtc-pc/#webidl-1143016005) nor [Chrome's implementation](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/peerconnection/rtc_data_channel.idl).

It is specified in https://w3c.github.io/webrtc-extensions/#rtcdatachannel-extensions.
It is implemented in Safari by reusing pre-existing mechanisms.
One reason why it might not be as difficult to implement as MST is that we restrict when RTCDataChannel can be transferred. This simplifies a lot the state handlings.
CropTarget is much simpler to transfer than RTCDataChannel since it has no changing state.

The fourth implementation that was brought to the discussion is the following:
1. CropTarget stores an identifier of the process where lives the element (or the document environment ID which is generated for each document before hand, so can be retrieved synchronously).
2. CropTarget stores a locally generated identifier of the element.
3. Element is identified by the pair of these two identifiers
4. CropTarget is serialised by serializing these two IDs.
5. When calling cropTo, the capture process identifies the element process through the first ID and gathers element information through the second ID.

Steps 1 to 4 can be done synchronously.

**Comment by alvestrand** [comment]
I don't understand the claim that "transferring is synchronous".
The "Message port post message steps" (https://html.spec.whatwg.org/multipage/web-messaging.html#message-port-post-message-steps) contains an "Add a task that runs the following steps to the port message queue of targetPort".

How can an algorithm that posts a task to a different task queue be considered synchronous?

**Comment by youennf** [comment]
> I would only design a synchronous interface if it was obvious that any reasonable implementation would never have a need for a step involving interactions with entities outside the renderer process hosting the API.

I do not think creating a cropTarget requires any "interactions with entities outside the renderer process", depending of course of your definition of interactions. Contrary to actually cropping a track for instance. Hence my conclusion that creating a CropTarget does not need to be asynchronous while cropTo should be asynchronous.

> How can an algorithm that posts a task to a different task queue be considered synchronous?

Sorry, this was misleading. What I tried to say is that the call to postMessage happens synchronously with the creation of the object. The whole postMessage algorithm is of course asynchronous.

@eladalon1983 mentioned several times that it has to be asynchronous between the time the web app wants to create the object and the time the web app wants to transfer the object. Performances and potential race conditions were mentioned.

The point I am making is that this issue has been solved multiple times and for objects more complex than a CropTarget.
This makes me think asynchronicity is not needed by reasonable UA implementations, they will just leverage existing UA patterns.

**Comment by jan-ivar** [comment]
> > 1. For any new API, we try to make it synchronous if we can.
> 
> I stopped agreeing at 1.

@alvestrand please read [§ 6.8. Use synchronous when appropriate](https://w3ctag.github.io/design-principles/#synchronous).

**Comment by eladalon1983** [comment]
I think there is a big difference between APIs that are **designed** for cross-render-process communication, and ones that don't. In this particular message I'd like us to think about **performance**. Namely, how long after calling `cropTo(target2)`, the cropping actually changes to `target2`. 
* With the current design, calling `cropTo()` is not impacted by how busy the main loop is on the render process hosting the target Element, because `cropTo()` only causes IPC with the browser process.
* In contrast, @youennf has suggested a design whereby `CropTarget` would encode (behind the scenes) which render process the Element is in. That means that `cropTo(token)` would validate the token through an IPC to that other render process - which means the IPC could be delayed by the other render process, e.g. if it is executing an especially long task.
  * This matters, because even async APIs should be fast. When changing to a new crop-target, the application wants the new crop to be applied ASAP.
  * This is NOT a problem for `produceCropTarget` because it happens long before the other process calls `cropTo()`; possibly even before `getDisplayMedia()` is called.

**Comment by youennf** [comment]
> * . That means that `cropTo(token)` would validate the token through an IPC to that other render process

When we discussed this, your objection was that it would be too costly to send IPC messages to each renderer process.
I thought we agreed that a single IPC message was ok.

> * which means the IPC could be delayed by the other render process

This is a solved problem: IPC messages are usually received in a specific run loop, then processed either in a dedicated run loop or the main loop. If main loop blocking is considered a perf issue in this particular context, a dedicated run loop will be used to gather the necessary element information from the provided ID (using a process global map protected by a lock for instance) and send back the information to the capturing process. This can be further optimised in common cases.

**Comment by eladalon1983** [comment]
> When we discussed this, your objection was that it would be too costly to send IPC messages to each renderer process.
I thought we agreed that a single IPC message was ok.

Could you please explain how a single IPC would work?

> This is a solved problem: IPC messages are usually received in a specific run loop, then processed either in a dedicated run loop or the main loop.

That is not a simple solution. I think performance is yet another argument in favor of solutions that don't involve communication between render processes. Engineering deals with trade-offs. The issue of making `produceCropTarget()` async, and the small loss in elegance involved therein, does not seem to me to merit the loss of (i) simplicity of implementation, (ii) strong security guarantees and (iii) performance.

**Comment by youennf** [comment]
> Could you please explain how a single IPC would work?

I was meaning a single IPC exchange.

> That is not a simple solution.

This is a widely used pattern. How is it not a simple solution?
Also, please bear in mind that this is an optional optimization.

> I think performance is yet another argument in favor of solutions that don't involve communication between render processes.

That is a UA implementation choice, this is orthogonal to go with sync or async.
Lazy and eager initialisations have both tradeoffs and are both doable.

> the small loss in elegance involved therein, does not seem to me to merit the loss of (i) simplicity of implementation, (ii) strong security guarantees and (iii) performance.

I disagree with reducing the evaluation to a loss in elegance.
About simplicity of implementation, this is of course debatable, I think we should assess it according existing UA patterns.
About security, there are AFAIUI no differences.
About performances, this seems orthogonal. UA will have to decide between a few additional milliseconds vs. the issues related to a central memory storage, or use hybrid solutions to get the best of both worlds.

**Comment by yoavweiss** [comment]
Apologies for the delay in responding here.

To reduce confusion (mine or otherwise), let me try to summarize my current understanding.

Based on @eladalon1983’s [previous comment](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1027843349) here, minting *[can fail](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/media/capture/crop_id_web_contents_helper.h;l=38;drc=f10df0040d17b75ac96491b1c009550e97778bc9)*. 

Such failures can happen if minting is done in the browser process or in the renderer process itself, e.g. if the developer is minting too many such tokens, and the implementation wants to impose limits on that. It seems reasonable that the API shape would allow for such failures to happen at minting time, and not just at crop time (where they’d be caught by a separate document, on a potentially separate origin).

Given the above, it seems to me that choosing a sync API shape for minting would significantly restrict implementations in their flexibility and force them to implement renderer-to-renderer communication channels which would require extra care for security purposes, and may suffer from performance issues, as pointed out [up-thread](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1128595737). Those potential performance issues would then cause *user* friction and frustration: We either introduce a delay when minting the tokens synchronously (as there’s a lock on the minting container, for off-thread IPC handling), or we introduce a delay when cropping (in case we need to establish communications with the other renderer’s busy main thread).

Reading through the [use sync when appropriate](https://w3ctag.github.io/design-principles/#synchronous) Web Platform Design Principle, it calls out the use of locks and IPCs as ones that *require async designs*. I think we all agree that a performant implementation of renderer-to-renderer communications would require a lock, at the very least.

Compounding the above with the [TAG’s opinion](https://github.com/w3ctag/meetings/blob/gh-pages/2022/telcons/04-25-minutes.md#media-capture-region) that an async design here doesn’t significantly decrease developer ergonomics, and the fact that this issue hasn’t come up as a hurdle to adoption as part of the [Origin Trial in Chrome](https://developer.chrome.com/origintrials/#/view_trial/2257429313219461121), I’m unconvinced that moving the API to be sync is warranted, as far as the priority of constituencies is concerned.

Starting out with an async API also seems more conservative, and a decision we could revert at a later point, if we deem it necessary. Implementations that don’t believe the above issues warrant asynchronicity could always return an immediately resolved Promise.

**Comment by youennf** [comment]
> Based on @eladalon1983’s [previous comment](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1027843349) here, minting _[can fail](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/media/capture/crop_id_web_contents_helper.h;l=38;drc=f10df0040d17b75ac96491b1c009550e97778bc9)_.

Thanks for bringing this, I missed that bit of information. A few related thoughts:
1. The produceCropTarget algorithm is currently not describing that it can fail. As it is, it cannot fail. If we think this is important, we should change the spec to describe when it fails so that implementations can have interop.
2. It is not clear how this 100 restriction kicks in. If it is per renderer process, it can be done inside the renderer process. If it is global to all renderer processes, I agree that some async IPC is the usual way to deal with this.
3. Can this 100 restriction be used as a cross-origin/cross-storage-partitioning messaging channel, say between two iframes living in the same process (out-of-process iframes is not a thing in all platforms) or between two iframes with the same origin but different top origins, or in case of same-process-but-cross-origin navigation? If this threat is real, we should probably change the design/implementation of such APIs. Having the failure at cropTo time is of course safe given capturer has access to capturee content.
4. Elad mentioned in the past that Chrome current implementation is doing some surface labelling processing at CropTarget creation time and that it might be better to move this to cropTo time. Would that alleviate the need for this restriction? Is Chrome's plan to keep that 100 restriction or to remove it at some point?
5. In general, APIs that can fail like this in odd cases is not really appealing, I doubt web developers will actually think of handling this error case and I am not sure what they can do to recover from this. Web applications will be better served with UA implementations that cannot fail. Web developers will have to deal with errors at cropTo time so it seems best to keep it the single error handling point.
6. Is Chrome having a similar restriction for MessagePort creation? Why is Chrome fine with synchronous MessagePort creation?

**Comment by eladalon1983** [comment]
> 1. The produceCropTarget algorithm is currently not describing that it can fail. As it is, it cannot fail. If we think this is important, we should change the spec to describe when it fails so that implementations can have interop.

Let's add it. We could specify that we allow the UA to fail for an implementation-specific reason. Or if we can't agree on that, we can document that some implementations do that.

> 2. It is not clear how this 100 restriction kicks in. If it is per renderer process, it can be done inside the renderer process. If it is global to all renderer processes, I agree that some async IPC is the usual way to deal with this.

Presently, in Chrome's implementation, it's global to all documents embedded within a given tab. Other browsers are free to implement with no limitation at all. Note that it's hard to come up with an app that would legitimately need more than 1-5 such tokens, so I don't think we'll have interop issues here if Chrome limits to 100 and Safari to 200.

> 3. Can this 100 restriction be used as a cross-origin/cross-storage-partitioning messaging channel, say between two iframes living in the same process (out-of-process iframes is not a thing in all platforms) or between two iframes with the same origin but different top origins, or in case of same-process-but-cross-origin navigation? If this threat is real, we should probably change the design/implementation of such APIs. Having the failure at cropTo time is of course safe given capturer has access to capturee content.

That's an interesting consideration. I can think of some mitigations, but when the time comes, I might need to tweak the Chrome implementation to address this. Having an async design increases everyone's flexibility in handling newly surfaced concerns such as this.

> 4. Elad mentioned in the past that Chrome current implementation is doing some surface labelling processing at CropTarget creation time and that it might be better to move this to cropTo time. Would that alleviate the need for this restriction? Is Chrome's plan to keep that 100 restriction or to remove it at some point?

Tagging at cropTo() time increases the time until the first frames are cropped and therefore reduces app performance - an important consideration for a highly ranked constituency. (IPC, start tagging in the render pipeline, etc.)

Tagging as soon as we call produceCropTarget() - our current implementation - means tagging elements incurs non-zero cost. It is good that we dissuade the application from making excessive calls without good reason.

> 5. In general, APIs that can fail like this in odd cases is not really appealing, I doubt web developers will actually think of handling this error case and I am not sure what they can do to recover from this.

They can modify their application to not tag more elements than strictly necessary.

> cropTo time so it seems best to keep it the single error handling point.

That error is in a different document.

> 6. Is Chrome having a similar restriction for MessagePort creation? Why is Chrome fine with synchronous MessagePort creation?

I was not around when MessagePort was designed and implemented. I cannot tell you whether compromises were made, or for what reason.

--

Consider, btw, that we're likely going to want to expand the API in the future, so that it would work with cropping a track derived of capturing ANOTHER tab. We don't need to get consensus on that just yet, but it's good to keep doors open. So what will we want then? Would it not be nicer if technical limitations are surfaced as errors on the tab that produced the error? Would it not be better that produceCropTarget() fail in the captured tab, rather than cropTo() in the capturing tab?

**Comment by youennf** [comment]
> They can modify their application to not tag more elements than strictly necessary.

That is not true in general, given how the web is authored today.
Web pages embed adds and do not control what adds will do (they might well do fingerprinting so will generate CropTargets)
Web pages are running in processes that run unrelated other pages, especially on small devices.

> That error is in a different document.

There is a communication channel between the two, otherwise there would be no CropTarget.

> Having an async design increases everyone's flexibility in handling newly surfaced concerns such as this.

The requirement for an async mechanism is based on an implementation that is showing big limitations.
It seems best to spend some time on improving the implementation as much as possible, then to look at whether async is actually required. I'll be happy to provide feedback to get great performances while keeping API sync.
Asking for async just in case it might be useful in the future does not seem like a good way to proceed.

> Tagging at cropTo() time increases the time until the first frames are cropped and therefore reduces app performance - an important consideration for a highly ranked constituency. (IPC, start tagging in the render pipeline, etc.)
> 
> Tagging as soon as we call produceCropTarget() - our current implementation - means tagging elements incurs non-zero cost. It is good that we dissuade the application from making excessive calls without good reason.

As said before, you can get the best of both worlds (no big repository/privacy issue, no latency in 99.99% cases) without too much issue. There is no need for an async API to achieve this.

> Would it not be nicer if technical limitations are surfaced as errors on the tab that produced the error? Would it not be better that produceCropTarget() fail in the captured tab, rather than cropTo() in the capturing tab?

I don't think so.
First, this creates potential privacy issues that we should avoid since produceCropTarget can be called without any permission. cropTo requires some permissions so no big deal exposing it at this point. We usually select designs that avoid those kind of issues hence the cropTo appeal.
Second, produceCropTarget will rarely fail. When it will fail, web applications will get broken. Typically a message will not be sent and the application will mysteriously hang.
cropTo will sometimes fail and web applications will probably have an error recovery mechanism that will be able to cover both cases. If we want to allow the web application to understand the issue, all we might need is to reject with an error indicating that the CropTarget is invalid/a new one is needed.

> > 11\. Is Chrome having a similar restriction for MessagePort creation? Why is Chrome fine with synchronous MessagePort creation?
> 
> I was not around when MessagePort was designed and implemented.

It is always good to look at existing APIs and existing UA patterns.
Deviating from these patterns is ok when we see clear shortcomings but we need to try applying how existing APIs/patterns deal with those same issues.

**Comment by jan-ivar** [comment]
@yoavweiss I'm hearing vastly differing claims about what [produceCropTarget](https://w3c.github.io/mediacapture-region/#dom-mediadevices-producecroptarget) needs to do that aren't supported by the spec: The spec says: _"Calling produceCropTarget on an [Element](https://dom.spec.whatwg.org/#element) of a supported type associates that [Element](https://dom.spec.whatwg.org/#element) with a [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget). This [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget) may be used as input to [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto)."_

That's literally all it says: create an association between an interface that cannot be serialized with one that can.

Clicking on [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget) confirms this: _"CropTarget is an intentionally empty, opaque identifier that exposes nothing. Its sole purpose is to be handed to [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto) as input."_ — Nothing about cropping, preparing for cropping, IPC, render processes, or any failures. It's infallible and serializable, that's all.

There's no reason this needs to be asynchronous. We've had experienced Google folks suggest element.id be used instead — a proposal that fell for other reasons — but being synchronous never came up. I agree with @youennf we need to go back to the drawing board if Chrome has vastly different ideas for this than what they've proposed for standardization. 

I can think of no reason why this needs to be async, I could literally polyfill CropTarget by minting a MessagePort, and it would work in 90% of cases (same doc, same origin, same site, which has similar performance characteristics AFAIK, just not true cross-doc).

Asynchronous API's come at a cost to web developers, as they turn JavaScript into a pre-emptive language. I recommend [Why coroutines won’t work on the web](http://calculist.org/blog/2011/12/14/why-coroutines-wont-work-on-the-web/) as required reading here. I can also talk at length about async functions not having been great for WebRTC.

**Comment by jan-ivar** [comment]
I should add (or I'll be called out on it) it goes on to say: _"The user agent MUST resolve p only after it has finished all the necessary internal propagation of state associated with the new [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget), at which point the user agent MUST be ready to receive the new [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget) as a valid parameter to [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto)."_ — but the sole requirement here seems to be that a newly created target MUST be accepted by cropTo, which is easily accomplished a number of ways given that any target must be postMessaged and cropTo is already async. The need for returning a promise for this is what has no consensus and is being challenged.

Also this second part is about of _how_ it does what it does, whereas my initial quote is more importantly about what it _does_, which doesn't match claims made here.

**Comment by yoavweiss** [comment]
> @yoavweiss I'm hearing vastly differing claims about what [produceCropTarget](https://w3c.github.io/mediacapture-region/#dom-mediadevices-producecroptarget) needs to do that aren't supported by the spec: The spec says: _"Calling produceCropTarget on an [Element](https://dom.spec.whatwg.org/#element) of a supported type associates that [Element](https://dom.spec.whatwg.org/#element) with a [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget). This [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget) may be used as input to [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto)."_
> 
> That's literally all it says: create an association between an interface that cannot be serialized with one that can.
> 
> Clicking on [CropTarget](https://w3c.github.io/mediacapture-region/#dom-croptarget) confirms this: _"CropTarget is an intentionally empty, opaque identifier that exposes nothing. Its sole purpose is to be handed to [cropTo](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto) as input."_ — Nothing about cropping, preparing for cropping, IPC, render processes, or any failures. It's infallible and serializable, that's all.

The current processing model seems indeed insufficient in describing what current implementations are doing (or planning to do). I sent https://github.com/w3c/mediacapture-region/pull/47 to clarify that part and make it (hopefully) more rigorous.


> There's no reason this needs to be asynchronous. We've had experienced Google folks suggest element.id be used instead — a proposal that fell for other reasons — but being synchronous never came up.

I [beg to differ](https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1125790108) on that last part.



> I agree with @youennf we need to go back to the drawing board if Chrome has vastly different ideas for this than what they've proposed for standardization

I don't believe that's the case. I provided spec clarifications at https://github.com/w3c/mediacapture-region/pull/47 in the hope that they'd help bridge our common understanding.

> I can think of no reason why this needs to be async

See [my comment above](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1129753450)


> Asynchronous API's come at a cost to web developers, as they turn JavaScript into a pre-emptive language.

Understood. At the same time, there's [general agreement](https://w3ctag.github.io/design-principles/#synchronous) that when locks or IPC calls are involved, async APIs are called for. I'd also like to stress out (again) that the TAG [did not find the burden significant](https://github.com/w3ctag/meetings/blob/gh-pages/2022/telcons/04-25-minutes.md#media-capture-region) in this particular case.

**Comment by youennf** [comment]
> I sent #47 to clarify that part and make it (hopefully) more rigorous.

Before doing this, I think we should first decide whether failing generation of CropTargets is something we want to expose to the Web and how. I filed https://github.com/w3c/mediacapture-region/issues/48 to dig into that. Can we continue this particular discussion there?

**Comment by eladalon1983** [comment]
> That is not true in general, given how the web is authored today.
Web pages embed adds and do not control what adds will do (they might well do fingerprinting so will generate CropTargets)

They should not embed abusive iframes. IIRC, you have yourself, @youennf, brought up a few objects of which browsers can only instantiate a limited number.

> There is a communication channel between the two, otherwise there would be no CropTarget.

* It is not necessarily two-way.
* It might be going through several embedders.
* It is not ergonomic to communicate the error back. Certainly less ergonomic than having production be async.

> no latency in 99.99% cases

There is latency between the time a frame is first tagged and the time it's rendered. If tagging happens ahead of time, cropTo() reduces latency 100% of the time.

> Second, produceCropTarget will rarely fail. When it will fail, web applications will get broken.

1. As I understand it, Safari and Firefox plan an implementation that imposes no limit on the number of tokens that can be minted. It is not mandated that produceCropTarget must fail. It would be an issue with Chrome's implementation.
2. I think the exact value of "rarely" is important here. If it's not too rare, it will be handled by applications. If it's exceedingly rare, like SHA-1 collisions, it's not a problem. Which specific value are you aiming at? Why are we worried?

> I filed #48 to dig into that. Can we continue this particular discussion there?

Gladly.

**Comment by yoavweiss** [comment]
> Regarding performance claims, it would seem faster to _not_ block postMessage, and let it happen in parallel with Chrome's implementation of produceCropTarget, and have cropTo deal with it not arriving in time with some backup strategy (try again a few times, or worst case do some IPC) before giving up entirely. This would seem the absolute fastest success path if it arrives in time, and worst case no slower than today where you're essentially serializing the two steps of generating the key and postMessaging it.

We have four options: browser-side sync, browser-side async, renderer-side sync and renderer-side async.
Let's examine the performance characteristics of each, shall we?

## renderer-side token production
Renderer side token production would require to establish renderer-to-renderer communications. If these communication channels are established on the main thread, sync failures are possible, but the downside is that a busy main-thread on SLIDE would mean that `cropTo` takes a long time to return.

If the communication channels are established in an auxiliary thread, that requires a lock when minting, which means we should go with an async interface, to avoid locking the main thread on this lock.

## Browser-side sync token production

A sync API and browser-side minting would not enable developers of SLIDE to know that the minting has failed (e.g. due to excessive minting of tokens on their behalf). The reason for that is that the communication to the browser process is async by nature, so the immediately returned token may or may not be a valid one.

In this case, we’d be relying on the VC side to notice failures when using `cropTo` and notify SLIDE’s developers in case it’s their fault.

Another issue is that with a sync API, we can have a race condition where `cropTo` is called by VC *before* the IPC from SLIDE has successfully completed.

Yet another case is VC calling `cropTo` with an expired token (e.g. the SLIDE document was detached).

How would an async `cropTo` with sync token minting respond to each one of those failure cases?

The async `cropTo` would need in this case to send an IPC to the browser process. If the token is in the token container, then all is well and `cropTo` is resolved.
If it’s not, it could be any one of the above cases:
* The token’s IPC hasn’t yet arrived
* The token was removed from the container
* The token is invalid because of bugs, it’s coming from a compromised renderer or any other reason

In order to properly handle the first case, the browser process would have to e.g. add an observer to token minting that would resolve the `cropTo` promise once the token has arrived.
But it would also need to have some sort of timeout that rejects `cropTo` after X seconds when the token doesn’t arrive.

That means that errors of using a removed or invalid token will take a long time to resolve. While one can argue that we could keep past tokens in memory, it’s unclear for how long that would be required. This would also have the undesired side-effect of increasing the long-lived browser process’ memory footprint (hurting users).

## Browser-side async token production

Finally, an async `produceCropTarget` would tell SLIDE when minting failed, and enable for shorter debug cycles.

It would also make sure that SLIDE doesn’t send the token to VC before the browser side is ready for it. That eliminates the “token hasn’t yet arrived” case, eliminating the need for timeouts and lengthy failures in case an invalid token is used by VC.

**Comment by youennf** [comment]
@yoavweiss, I think there are other approaches that sounds better, for instance:
- When creating a CropTarget, register it in renderer process. Send an IPC to the capturer process doing the capture. These are synchronous tasks. No need to wait for capturer process answer.
- Capturer process stores a limited set of CropTarget informations based on received IPC, handled using round robin say.
- When capturer is asked to crop, capturer looks at its limited set of CropTarget. In 99.99% of the cases, capturer will be able to start right away without waiting for renderer
- If no CropTarget is available, send an IPC to renderer which will give back the necessary information to do cropping.

This approach removes the capturer process memory capping requirements/privacy issues, removes the need for web pages to have to deal with CropTarget failures, leads to maximum performances in almost all reasonable cases and is reasonably simple.

**Comment by eladalon1983** [comment]
> handled using round robin say

I am firmly opposed to this too. This has been discussed multiple times. Using this scheme means that some tokens are zombie-tokens, meaningless but undetectable except via cropTo().

I think we're rehashing the same discussions multiple times.

**Comment by eladalon1983** [comment]
> I do not know how Chrome is implemented by I would guess that there is a privileged process that is doing the compositing, might have access to window server and so on, I am talking about this process. Basically this is the process where lives the getDisplayMedia track source. Isn't there such a thing in Chrome implementation? The browser process maybe?

There's a non-trivial system of interactions between the render processes, browser process and GPU process. It's not clear to me how your proposal maps onto our model; I am very confused by [your proposal](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1131856421) and its reference to such things as "synchronous IPC".

Equally important, it's not clear to me why we can't go with a general, flexible solution that can be easily accommodated by any reasonable user-agent architecture. I am [still waiting](https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1126813025) for an explanation of how this complexity provides a service to higher constituencies. Barring such compelling motivation, what's the use?

> There are no zombie tokens. If a CropTarget is kicked out of the CropTarget capturer process cache, it will still be in the renderer process and capturer process will get it through an IPC exchange. The size of the CropTarget cache can be tailored at will since it is just an optimization and has no impact on JS surfacing functionality.

So now we have a round-robin of cached CropTargets that can be resolved immediately, but if one is ejected we do render-to-render communication to figure out if it's a legitimate CropTarget or not? And we set up this potential channel for communication ahead of time, just in case? And this channel of communication follows the capturer around, and supposedly we only have one, and supposedly it's stationary? And all of this wonderful complexity is in the service of...?

**Comment by jan-ivar** [comment]
A test in Chrome with https://jsfiddle.net/jib1/yunghz8f/ (on macOS in my case) reveals the following: On a fresh restart:
```
Time to crop: 83.00
Time to play: 537.20
```
...and on repeat calls more like:
```
Time to crop: 3.30
Time to play: 486.60
```
Numbers fluctuate a bit, but cropping always succeeds with a substantial margin ahead of the video playing.

I then did an extremely rough polyfill of `element.cropTarget` in https://jsfiddle.net/jib1/wfnv9o3q/ solely for the purpose of timing things, and got numbers like this:
```
Time to crop: 2.90
Time to play: 509.60
```
I hope this puts to bed any performance concerns, but if someone wants to do a cross-origin version feel free.

The fiddles also show some bugs in Chrome you may wish to iron out (the blue outline is way off)

**Comment by eladalon1983** [comment]
@jan-ivar, I don't understand your methodology for measuring performance.
* How am I going to extrapolate the numbers from your machine or mine, to the expected numbers for users on lower-end hardware, or with busy CPUs?
* What are you measuring? Time to the resolution of the Promise returned by cropTo()? The [CL](https://chromium-review.googlesource.com/c/chromium/src/+/3497676) for that is still pending submission. I have not had the time to finalize the tests because I've been busy debating specs. Without that CL, you're not measuring what you think. You're just measuring IPCs between the render process and the browser process.
* Even after that specific CL is submitted, this fiddle will measure when the last pre-crop frame was produced, not when the new post-crop frame is produced - which is the more interesting number. Because...
* Any measurement that does not result in a latency that falls below half the rate at which frames are drawn to the screen, is suspect. Because frames are captured when drawn. (Given our current implementation.)

> The fiddles also show some bugs in Chrome you may wish to iron out (the blue outline is way off)

Thanks for pointing that out. We'll definitely look into it.

**Comment by yoavweiss** [comment]
> Asynchronous API's come at a cost to web developers, as they turn JavaScript into a pre-emptive language.

How so?

> I recommend [Why coroutines won’t work on the web](http://calculist.org/blog/2011/12/14/why-coroutines-wont-work-on-the-web/) as required reading here.

Reading through this, I'm failing to understand the relevance. Are you suggesting that Promises are the equivalent of co-routines?

**Comment by eladalon1983** [comment]
> > The fiddles also show some bugs in Chrome you may wish to iron out (the blue outline is way off)
> 
> FYI I've filed a Chromium bug at https://bugs.chromium.org/p/chromium/issues/detail?id=1327560 to track this. Thank you @jan-ivar

Thanks you too, @beaufortfrancois. It was past midnight in my timezone, so I left filing this formally for today. It was a very pleasant surprise to see you've done this for me.

**Comment by eladalon1983** [comment]
> > And all of this wonderful complexity is in the service of...?
> 
> @eladalon1983, when looking at this API...

Thank you for this message, @youennf. I know all messages take effort, and I value your time. Alas, having read this message three times, I still cannot find in it an answer to my original question. Namely - what is this [incredible complexity](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132095029) serving? I continue to await an answer. I shall now answer your question, but I'd appreciate it if you did not forget to answer mine.

> Is it loosing 1 or 2 frames or one second of frames?

Being slower to start, stop, or change the crop target, is unfortunate. An application that decide to change the crop-target will often pause the track until it is informed that all subsequent frames are cropped to the new target. This can be noticeable to both the local user as well as remote users. This visible disruption should be minimzed.

> And start cropping + store them in a round robin cache to optimise the regular cases. Caching is a well known optimisation tool so I think we can qualify this as a simple approach.

A render-to-render communication channel would have to be produced to support the case of a cahce-miss. A case you expect to happen [0.01% of the time](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1129888809). But this backup would have to be set up every time. That complexity is not warranted. (And please do NOT read this message as an acknowledgement that any of the rest of your proposed solution works.)

**Comment by eladalon1983** [comment]
Gentlemen, I have a simple proposal.
```webidl
  Promise<void> cropTo(Element);
  Promise<void> cropTo(CropTarget);
```

1. Same document:
```js
  // Note that *only* cropTo() is async here, which is
  // within consensus.
  await track.cropTo(document.getElementById('id'));
```

2. Multiple documents:
```js
  function totallySyncFunction() {
    setTimeout(async () => {
      await getCropTargetAndSendItToCapturer(element);
    }, 0);
  }

  async function getCropTargetAndSendItToCapturer(element) {
    await CropTarget.fromElement(element);
    otherDocument.postMessage(...);
  }
```

Note that, because **receiving** the message is asynchronous, the receiver would not observe a difference. This seems completely palatable for a generally synchronous sender.

Wdyt?

**Comment by jan-ivar** [comment]
> ... I have a simple proposal.

@eladalon1983 Isn't that a rehash of https://github.com/w3c/mediacapture-region/issues/19, and [already answered](https://github.com/w3c/mediacapture-region/issues/19#issuecomment-1126946787) there (special-casing != KISS)? Let's keep that there.

> ```js
>   setTimeout(async () => {
>     await getCropTargetAndSendItToCapturer(element);
>   }, 0);
> ```

As a side-comment: I don't know how you can credibly argue about performance and add 4 milliseconds of delay for no reason whatsoever. Check out [postMessage is ~100 times more frequent than setTimeout(0)](http://jsfiddle.net/jib1/y1e4rg0d/).

**Comment by jan-ivar** [comment]
> > Asynchronous API's come at a cost to web developers, as they turn JavaScript into a pre-emptive language.
> 
> How so?

@yoavweiss Everywhere a web developer has written `await` they're preempted, which means all their state assumptions made right before need to be reevaluated. This is a concrete cost to web developers, and can be a source of hard-to-diagnose races and bugs.

Additionally, once a function or method is made asynchronous, everything that calls it is forced into being asynchronous as well, so the problem spreads like wildfire.

The article I linked to explains this problem in some detail with examples, so I'm not sure what's unclear. Hope this helps. If not, I trust sure there are Google colleagues that can explain this better than I can.

> > I recommend [Why coroutines won’t work on the web](http://calculist.org/blog/2011/12/14/why-coroutines-wont-work-on-the-web/) as required reading here.
> 
> Reading through this, I'm failing to understand the relevance. Are you suggesting that Promises are the equivalent of co-routines?

No, the article was written in 2011 when Promises existed, but `async`/`await` didn't. At the time there was pressure from web developers to solve what that ultimately ended up solving a few years later. But how to do it wasn't yet clear.

### Ever questioned why the `await` keyword is needed, or why only `async` functions can use it?

This is that article. A the time it wasn't clear why those restrictions were needed. But take away either of those requirements, and _then_ you have "co-routines" (the author's term for every function you call being able to preempt you in a single-threaded environment).

The gem of the article (besides describing the preemption problem and why JS doesn't need locks), was recognizing that [generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) didn't have this problem, because the `yield` keyword was explicit. This meant JS developers could tell where preemption would happen and where it couldn't, so they could take appropriate steps. As you may know, `async`/`await` in JS ended up being [built on](https://stackoverflow.com/questions/36196608/difference-between-async-await-and-es6-yield-with-generators) generators.

The misguided _make-everything-async-there's-no-downside_ crowd don't seem to understand this or care, that the logical end of what they're advocating (`await` everything!¹) quickly degrades JS into this pre-emptive (co-routine) language again. I'm sure they would try to invent locks in JS to solve it. Thankfully, these views are more rare these days than they were in 2011, which is why I had to go back that far for sourcing.

---
<sub>1. I'm contrasting extremes to aid explanation here. Clearly one more async method isn't the problem, nor does the problem doesn't only arise at the extreme. But my aim here was to clarify why the design rules we have matter, and mustn't be violated except for reasons we can all agree are sufficient.</sub>

**Comment by youennf** [comment]
> Namely - what is this [incredible complexity](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132095029) serving?

What I tried to say was that I do not see a big difference of UA implementation complexity/achievable performances between what Chrome implements and the proposed alternatives. The big difference I see is that Chrome current implementation puts some of the complexity on web pages in the name of this optimisation. In general, I think we should favour optimisations that are transparent to web pages.

> An application that decide to change the crop-target will often pause the track until it is informed that all subsequent frames are cropped to the new target. This can be noticeable to both the local user as well as remote users. This visible disruption should be minimzed.

I do not really understand in which conditions the application will pause the track, which I interpret as stop displaying/sharing it between the time cropTo is called and the time cropTo promise is resolved.

I can see a few related scenarios:

1. The web page wants to call getDisplayMedia to capture itself, apply cropping, then start displaying/sending the cropped track.
One natural approach here is to make just one getDisplayMedia call (something like `getDisplayMedia({ selfBrowserSurface: include, video : { cropTarget : myCropTarget } }`)).
This should simplify the web page workflow and allow some UA optimisations to get the smallest achievable latency.
Is it something that we should push to a spec as a way to reach consensus?

2. The web page is already capturing the tab and is displaying/sharing it. User decides to apply cropping through the web page UI.
In that case, calling cropTo after getDisplayMedia is the only way to implement this scenario.
This seems similar to when user requests to change camera (going from front to back camera say). I do not think web pages will pause the track in that scenario.
There is of course a need for cropTo to be fast enough so that the UI is responsive and that there is no excessive latency between user click and actual cropped frames to appear on screen.
In that case, it does not make sense to me for the web page to pause the track.

**Comment by eladalon1983** [comment]
> As a side-comment: I don't know how you can credibly argue about performance and add 4 milliseconds of delay for no reason whatsoever. Check out [postMessage is ~100 times more frequent than setTimeout(0)](http://jsfiddle.net/jib1/y1e4rg0d/).

The token can be posted while the user is interacting with getDisplayMedia's dialog, so any latency introduced by minting a token becomes zero. This wonderful fact is courtesy of the produceCropTarget/cropTo separation as two separate async APIs.

**Comment by eladalon1983** [comment]
> What I tried to say was that I do not see a big difference of UA implementation complexity/achievable performances between what Chrome implements and the proposed alternatives.

I humbly disagree. I believe I have argued the point.

> I do not really understand in which conditions the application will pause the track, which I interpret as stop displaying/sharing it between the time cropTo is called and the time cropTo promise is resolved.

You gave two examples. You missed a third - multiple targets. Assume the application has two sub-sections that the user can switch between sharing by clicking an in-content button. If I were to implement this app, I'd see a click from the user as telling me: (i) stop capturing (and/or transmitting-remotely) the old target's content and (ii) start capturing the new target's content. I'd pause the track until (ii) is completed. Depending on the nature of the app; it would only matter for some.

**Comment by youennf** [comment]
> I humbly disagree. I believe I have argued the point.

It seems we are stuck on this complexity assessment.

I hope we agree though that Chrome's current implementation is putting additional complexity on the web page shoulders for the sake of this optimisation.

> You gave two examples.

I gave examples where pause is not required. Your third example is a generalisation of the second example, where pause is not required, at least to the apps I can think of.

> If I were to implement this app, I'd see a click from the user as telling me: (i) stop capturing (and/or transmitting-remotely) the old target's content and (ii) start capturing the new target's content. I'd pause the track until (ii) is completed.

I do not see what would drive user to expect pausing here.
I would think VC apps tend to not follow that pattern, they prefer a seamless transition.
replaceTrack is suited for that seamless transition to go from old track to new track.
Also replaceTrack is a chained operation: it was thought fine to delay a bit switching the sources.
This probably mean that some limited switch latency is fine for most WebRTC applications.

If the app is so keen to do that kind of optimisations, they can do it on their side in a faster way than any UA-side optimizations:
- When getting getDisplayMedia track, clone the track and call cropTo on the cloned track.
- When user clicks, switch to either the original or the clone, synchronously. No need to do any pausing.

> This depends on the the app; it would only matter to some.

I still do not understand which apps would actually want to do what you described, and for which reasons.

**Comment by eladalon1983** [comment]
Holding multiple tracks but only using one at any given time, may incur performance costs, or be subject to implementation-specific practical limitations. However, I don't think this is core to our discussion, so I suggest we drop this particular sub-topic. It should be pretty clear that applications calling `cropTo()` prefer if the call has an effect sooner rather than later.

**Comment by youennf** [comment]
> It should be pretty clear that applications calling `cropTo()` prefer if the call has an effect sooner rather than later.

Not at all cost, see replaceTrack precedent for instance.
All transparent optimisations are fine. None transparent optimisations like the one we are discussing here should be heavily justified.

**Comment by jan-ivar** [comment]
> > As a side-comment: I don't know how you can credibly argue about performance and add 4 milliseconds of delay for no reason whatsoever. Check out [postMessage is ~100 times more frequent than setTimeout(0)](http://jsfiddle.net/jib1/y1e4rg0d/).
> 
> The token can be posted while the user is interacting with getDisplayMedia's dialog. Or even before getDisplayMedia is called. Any latency introduced by minting OR posting a token therefore becomes **zero**. 

@eladalon1983 I'm sorry are you still arguing for `setTimeout`? Do you understand its purpose is specifically to induce delay? If you meant to simply not wait for completion, just remove the `await` instead (but let's move this idea to #19):
```js
  function totallySyncFunction() {
    getCropTargetAndSendItToCapturer(element).then(() => {});
  }
```

Back to performance. I've shown a sync minting API in JS is [actually faster](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132232768) because otherwise [you're essentially serializing the two steps of generating the key and postMessaging it](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1130846621). This seems to undermine your hypothesis that your design is faster. Do you disagree?

@yoavweiss's response suggests your design only [eliminates the “token hasn’t yet arrived” case, eliminating the need for timeouts and lengthy failures in case an invalid token is used by VC](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1131846988). So your design sacrifices optimal success path of `cropTo` in order to optimize for its failure path? Am I missing something?

And in your response here you're finally claiming performance doesn't matter? Color me confused. Or are you criticizing my measurement fiddle for not demonstrating *when* the performance I measure would matter? If so, it's true I struggled to imagine an example when this would matter, but that would seem to be on you to produce. You've mentioned multiple targets... please show how the shape of `produceCropTarget` would have any bearing in that case. It would presumably have to start with IPC from a decision point, just like my measurement fiddle (which only starts measuring _post_-gDM success btw), and presumably open to the same criticisms that decisions could have started even earlier. Let's see it.

> > The misguided make-everything-async-there's-no-downside crowd don't seem to understand this or care, that the logical end of what they're advocating (await everything!¹) quickly degrades JS into this preemptive (co-routine) language again. I'm sure they would try to invent locks in JS to solve it. Thankfully, these views are more rare these days than they were in 2011, which is why I had to go back that far for sourcing.
> 
> Nobody here has ever advocated for making **everything** async, or re-inventing locks. Let's stick to the arguments presented.

You seem to be in the "no-downside crowd", claiming [there's no downsides](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132690689) in the face of me clearly explaining them (having to reevaluate state assumptions, fueling the proliferation of unnecessary async, non-idiomatic JS confusing web devs into thinking the API does more than it does, and finally, performance).

Let me blow up my footnote that spelled out the relevant point it seems you flat-out missed:

1. I'm contrasting extremes to aid explanation here. Clearly one more async method isn't the problem, nor does the problem only arise at the extreme. But my aim here was to clarify why the design rules we have matter, and mustn't be violated except for reasons we can all agree are sufficient.

**Comment by eladalon1983** [comment]
> @eladalon1983 I'm sorry are you still arguing for setTimeout? Do you understand its purpose is specifically to induce delay? If you meant to simply not wait for completion, just remove the await instead (but let's move this idea to https://github.com/w3c/mediacapture-region/issues/19):

Messages only make sense in their proper context.
* **What I did:** I described the technic by which a synchronous function could participate in cropping, even if the developer did not wish to make that function asynchronous.
* **What I did NOT do:** Advocate for using setTimeout() in the general case of cropping.

**Comment by eladalon1983** [comment]
> @eladalon1983 I'm sorry are you still arguing for setTimeout? Do you understand its purpose is specifically to induce delay? If you meant to simply not wait for completion, just remove the await instead (but let's move this idea to https://github.com/w3c/mediacapture-region/issues/19):

Messages only make sense in their proper context.
* **What I did:** I described a technique by which a synchronous function could participate in cropping, even if the developer did not wish to make that function asynchronous.
* **What I did NOT do:** Advocate for using setTimeout() in the general case of cropping.

> Back to performance. I've shown a sync minting API in JS is [actually faster](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132232768) because otherwise [you're essentially serializing the two steps of generating the key and postMessaging it](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1130846621). This seems to undermine your hypothesis that your design is faster. Do you disagree?

Did you read [my response](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132272361)?

**Comment by eladalon1983** [comment]
> in the face of me clearly explaining them (having to reevaluate state assumptions, fueling the proliferation of unnecessary async, non-idiomatic JS confusing web devs into thinking the API does more than it does, and finally, performance)

I've asked you for the downsides. To your mention of these specific downsides, I [answered](https://github.com/w3c/mediacapture-region/issues/17#issuecomment-1132746253):
1. Let's use `cropTarget(Element)` in the same document - no **new** asynchronicity.
2. Let's use the other API when posting to another document - posting a message is asynchronous **anyway**. And if your **sender** is synchronous - which is unlikely - let it just painlessly use `setTimeout(async...)`. It's barely noticeable to the sender, and completely invisible to the receiver.

**Comment by azizj1** [comment]
Hello there. Aziz Javed from Google Meet here, working with [Lindsay](https://github.com/w3c/mediacapture-region/issues/31) from Google Editors. We've already built a very significant feature based on this API's origin trial. I'm happy to inform y'all that we were not at all inconvenienced by the async nature of produceCropId. Exposure on MediaDevices vs. on CropTarget is likewise not at all a concern for us; either approach works. Generally speaking, all alternatives debated here are perfectly reasonable for us, so long as they're performant. We'd favor whichever API shape would allow this feature to be shipped in as many browsers as possible, as early as possible.

**Comment by jan-ivar** [comment]
> We'd favor whichever API shape would allow this feature to be shipped in as many browsers as possible, as early as possible.

@azizj1 glad to hear this! I believe we've already resolved exposure on MediaDevices, so for me that API shape is what @youennf proposed in https://github.com/w3c/mediacapture-region/issues/11#issuecomment-1129883808:
```js
new CropTarget(element)
```
Sorry this is taking so long to get to the bottom of.

**Comment by youennf** [comment]
Quickly discussed at today's editor's call, hopefully we can discuss and resolve this issue at next WG interim (next Tuesday).

**Comment by eladalon1983** [comment]
Repeating from elsewhere for better visibility:
I am not sure I'll be able to attend. I regret that the intention to discuss https://github.com/w3c/mediacapture-region/issues/17 was announced only today, or I would have cleared my schedule ahead of time. (As of earlier today, no agenda items were announced.) I'll try to shuffle around my other commitments and make it. But if I cannot, I suggest that it might be more productive to discuss Region Capture during an interim meeting which I can attend. For your consideration.

CC @aboba and @dontcallmedom

**Comment by eladalon1983** [comment]
For those interested, the rescheduled meeting will take place 2022-06-23. Seem more details [here](https://www.w3.org/2011/04/webrtc/wiki/June_23_2022).

**Comment by aboba** [comment]
After looking over the slides and listening to the arguments, my take is the following: 

1. [The W3C design principles]([ttps://www.w3.org/TR/design-principles/#synchronous](https://www.w3.org/TR/design-principles/#synchronous)) cite interprocess communications (IPC) as a potential reason to choose async over sync, so they don't provide much in the way of definitive guidance here. 
2. There is a good argument that `cropTo()` needs to be async. 
3. The concerns about resource exhaustion attacks against `CropTarget()` relate more to a potential implementation issue than to the async vs. sync question.  That is, resource exhaustion attacks appear addressable even if `CropTarget()` remains async. For example, resource allocation could be moved from `CropTarget()` to `cropTo()`, since the latter is async. 
4. As Tim Panton pointed out, the only existing implementation uses async and the developers do not believe that a sync approach would be feasible for them.  On the other hand, those who favor sync could adopt the async approach and if desired, move resource allocation and other potentially blocking operations to `cropTo()`, returning much sooner.  So it would appear that all implementers should be able to "live with" async, whereas some implementers claim they cannot live with the sync approach.

**Comment by yoavweiss** [comment]
> It says to use sync if the "[API ... will not be _blocked_ by ... inter-process communication.](https://www.w3.org/TR/design-principles/#synchronous)" (emphasis on _blocked_)

You missed "locks" inside the elipsis



> here's no reason for JS to block on it since it cannot fail

It [can though](https://github.com/w3c/mediacapture-region/issues/48)

> Needs for failure are discussed and disputed in #48.

"disputed" != "settled". There's still an ongoing discussion.

> Exactly. [in Chrome: mint a dud `CropTarget` that always fails in `cropTo` instead]

That would create a significantly worse developer experience, as failures would be disconnected from the code/origin/party that causes them.

**Comment by jan-ivar** [comment]
Are there "locks" involved other than IPC? The FPWD says it cannot fail is what I meant. I agree about settling #48 first.

Who "caused" the (global?) resource error seems important only if they can do anything about it. We have no API for _relinquishing_ CropTargets, only minting more, so any recovery would violate [§ 5.3. Don’t expose garbage collection](https://w3ctag.github.io/design-principles/#js-gc).

**Comment by jan-ivar** [comment]
> I was suggesting that such a solution also requires an async API, to avoid the lock blocking the main thread.

I don't think it does. Any API can trigger [in parallel](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel) work. Returning a promise is only valuable if the caller needs the result.

If we "avoid blocking" then it says to use sync (if the "[API ... will not be blocked by ... a lock.](https://www.w3.org/TR/design-principles/#synchronous)")

**Comment by aboba** [comment]
If `CropTarget()` is vulnerable to resource exhaustion attacks wouldn't that imply that`CropTarget()' could fail (e.g. due to lack of resources)?  

Moving some or all resource allocation to `CropTo()` could result in the resource allocation arising later (in `CropTo()` instead of `CropTarget()`).  So depending on where the resource allocation is done, couldn't either or both of `CropTarget()` and `cropTo()` fail,  for similar reasons? 

Recently I was asked to investigate a bug in a sync API that is supposed to always return immediately.  In the most common use cases, it did return immediately (< .1 ms), no matter how much load was placed on it.   However, in a particular (reproducible) set of circumstances, it blocks for 30-50 ms. The cause appears to be a mixture of resource (de)-allocation and blocking IPC. It's probably a bug not a feature, but sometimes sync APIs block for unanticipated reasons.

**Comment by jan-ivar** [comment]
> If `CropTarget()` is vulnerable to resource exhaustion attacks wouldn't that imply that`CropTarget()' could fail (e.g. due to lack of resources)?  

Causality runs the other way: Letting CropTargets fail allows for implementations vulnerable to exhaustion attacks. Not doing so, doesn't.

A sensible implementation should be invulnerable to resource exhaustion attacks, by simply not tying resources to a token so easily created by anyone.

> Moving some or all resource allocation to `CropTo()` could result in the resource allocation arising later (in `CropTo()` instead of `CropTarget()`).  So depending on where the resource allocation is done, couldn't either or both of `CropTarget()` and `cropTo()` fail,  for similar reasons? 

What resource allocation is needed? A sensible `cropTo` implementation can use IPC to find the element it's supposed to crop to without consuming any finite resources. Also, `cropTo` is behind `getDisplayMedia` permission.

Chrome has implemented a neat but premature optimization, and refuse to implement the fallback needed to hide the resource exhaustion they’ve exposed themselves to.

I don't find the idea that creating a `{}` crop target will ever take appreciable time convincing.

**Comment by eladalon1983** [comment]
> Chrome has implemented a neat but premature optimization, and refuse to implement the fallback needed to hide the resource exhaustion they’ve exposed themselves to.

The spec does not compel anyone to replicate our alleged neat mistakes.

**Comment by youennf** [comment]
The issue is not really about other UA implementations but about web developer impact.
For instance, given Chrome is globally capping the number of CropTarget due to resource exhaustion risks, web developers might want to cautiously create CropTargets, typically only once they know the web page is being captured.
This goes against the idea of a web site proactively creating some CropTargets.
This might also have an impact on future additions to this API.

It would be good to understand whether Chrome has a plan to fix this global resource exhaustion issue.
As I understand it, the main problem seems to be implementation cost.

I already said that but working on https://github.com/w3c/mediacapture-region/issues/48 may allow to make progress on this particular issue.

**Comment by eladalon1983** [comment]
> The issue is not really about other UA implementations but about web developer impact. For instance, given Chrome is globally capping the number of CropTarget due to resource exhaustion risks, web developers might want to cautiously create CropTargets, typically only once they know the web page is being captured. This goes against the idea of a web site proactively creating some CropTargets. This might also have an impact on future additions to this API.

As previously explained, the global limit is an artifact of the current implementation, and can be changed in a number of ways. The minimal step forward is to make the limitation per-iframe rather than per-tab; that much is likely quite simple, and I doubt more would be needed.


> It would be good to understand whether Chrome has a plan to fix this global resource exhaustion issue. As I understand it, the main problem seems to be implementation cost.

I'll be happy to share more about my plans for prioritizing work in Chrome if you would be interested in likewise sharing Apple's timelines for implementing this feature. Given the high engagement, may I read that both Mozilla and Apple are prioritizing this work highly and intend to accomplish it in the near future?
